### PR TITLE
Node list feature

### LINKED ISSUE #35: How to add nodes that are not fit for ansible
In the current example we always assume that we can connect to the nodes we want to monitor to gather facts that are required to create the node in OpenNMS.
Another scenario could be to just add a list of switches, appliances or IP addresses that we want to monitor.

### PR DESCRIPTION
**Pull request by mfuhrmann**
fix #35

Can be tested with `ansible-playbook -i inventory 03-switches.yml`.

### CODE REVIEW

File: ansible/03-switches.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+# file: horizon-provision.yaml
+
+- hosts: switches
+  gather_facts: false
```
Comment (mfuhrmann) [inline ansible/03-switches.yml:5]: `false` because we can't use ansible to ssh into it.

File: ansible/group_vars/switches/vars.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+env: production
+ansible_connection: local
```
Comment (mfuhrmann) [inline ansible/group_vars/switches/vars.yml:3]: This lets Ansible know that it should not connect to the node.

File: ansible/inventory/03-switches.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
```
Comment (mfuhrmann) [inline ansible/inventory/03-switches.yml:1]: This is the inventory list. Not sure if we could get it smaller when it is about the notation.

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 4]
```diff
@@ -22,7 +22,7 @@
       Content-Type: "application/xml"
     body: "{{lookup('template','requisitions_xml.j2')}}"
   no_log: true
-  register: requisitions_created
+  register: requisition_created
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:25]: Fixed a typo. We never wrote for the `registers` in plural.

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 5]
```diff
@@ -36,13 +36,13 @@
       Content-Type: "application/xml"
     body: "{{lookup('template','foreignSources_xml.j2')}}"
   no_log: true
-  register: requisitions_created
+  register: requisition_created
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:39]: Fixed a typo. We never wrote for the registers in plural.

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 6]
```diff
@@ -36,13 +36,13 @@
...
   when: requisition_exists
   with_dict: "{{ onms_policies }}"
 
 - name: "Verify if nodes exist in requisitions"
   ansible.builtin.uri:
-    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] }}"
+    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}"
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:45]: `machine_id` is not available on nodes that don't get facts. We take the `inventory_hostname` here which is defined in the inventory.

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 7]
```diff
@@ -69,7 +69,7 @@
 
 - name: "Verify if monitored services exists"
   ansible.builtin.uri:
-    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] }}/{{ onms_hzn.interfaces_api }}/{{ ansible_host }}/services/{{ item }}"
+    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}/{{ onms_hzn.interfaces_api }}/{{ ansible_host }}/services/{{ item }}"
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:72]: https://github.com/opennms-forge/ansible-provisioning/pull/36/files#r1284402812

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 8]
```diff
@@ -83,15 +83,16 @@
 
 - name: "Create monitored services"
   ansible.builtin.uri:
-    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] }}/{{ onms_hzn.interfaces_api }}/{{ ansible_host }}/services/"
+    url: "{{ onms_hzn.base_rest_url }}/{{ onms_hzn.requisitions_api }}/{{ onms_requisition }}/{{ onms_hzn.node_api }}/{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}/{{ onms_hzn.interfaces_api }}/{{ ansible_host }}/services/"
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:86]: https://github.com/opennms-forge/ansible-provisioning/pull/36/files#r1284402812

File: ansible/roles/horizon-provision/tasks/main.yml [hunk 9]
```diff
@@ -83,15 +83,16 @@
...
     user: "{{ onms_hzn.user }}"
     password: "{{ onms_hzn.password }}"
     method: POST
     status_code: 202
     headers:
       Content-Type: "application/xml"
     body: "{{ lookup('template','services_xml.j2') }}"
-  no_log: false
+  no_log: true
+  ignore_errors: true
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/tasks/main.yml:95]: I think we missed this one before, don't we?

File: ansible/roles/horizon-provision/templates/node_xml.j2 [hunk 10]
```diff
@@ -1,10 +1,16 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
 <!-- {{ ansible_managed }} -->
-<node location="{{ onms_location }}" building="{{ onms_requisition }}" foreign-id="{{ ansible_facts['machine_id'] }}" node-label="{{ ansible_facts['hostname'] }}">
+<node location="{{ onms_location }}" building="{{ onms_requisition }}" foreign-id="{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}" node-label="{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}">
   <interface descr="disc-if" ip-addr="{{ ansible_host }}" managed="true" status="1" snmp-primary="P"/>
   <category name="{{ env }}"/>
+  {% if ansible_facts['product_serial'] is defined %}
```
Comment (mfuhrmann) [inline ansible/roles/horizon-provision/templates/node_xml.j2:6]: Just adding them when the facts are available.

File: ansible/roles/horizon-provision/templates/node_xml.j2 [hunk 11]
```diff
@@ -1,10 +1,16 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
 <!-- {{ ansible_managed }} -->
-<node location="{{ onms_location }}" building="{{ onms_requisition }}" foreign-id="{{ ansible_facts['machine_id'] }}" node-label="{{ ansible_facts['hostname'] }}">
+<node location="{{ onms_location }}" building="{{ onms_requisition }}" foreign-id="{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}" node-label="{{ ansible_facts['machine_id'] if ansible_facts['machine_id'] is defined else inventory_hostname }}">
```
[the remark this answers is not in the document]
Comment (indigo423) [reply ansible/roles/horizon-provision/templates/node_xml.j2:3]: Maybe we can find a way to use a variable for foreign-id and node-label, so we don't have to do this 'if defined...` not all the time?

File: horizon/docker-compose.yaml [hunk 12]
```diff
@@ -91,3 +91,15 @@ services:
       opennms:
         ipv4_address: 172.16.238.13
     privileged: true # Required for /usr/sbin/init
+
```
[the remark this answers is not in the document]
Comment (indigo423) [reply horizon/docker-compose.yaml:94]: I like the idea, I think we can keep it with the fake switch here to explain the idea.

**Comment by indigo423** [comment]
@mfuhrmann while I was deploying the fake switches I got this error message:

```
TASK [Gathering Facts] ***********************************************************************************************************************************************************************************
fatal: [switch1]: FAILED! => {"ansible_facts": {}, "changed": false, "failed_modules": {"ansible.legacy.setup": {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python3"}, "failed": true, "
module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}}, "msg": "The following modules failed to execute: ansible
.legacy.setup\n"}                                                                                                                                                                                         
fatal: [switch2]: FAILED! => {"ansible_facts": {}, "changed": false, "failed_modules": {"ansible.legacy.setup": {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python3"}, "failed": true, "
module_stderr": "sudo: a password is required\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}}, "msg": "The following modules failed to execute: ansible
.legacy.setup\n"}
```

Is there anything I need to do here?

**Comment by indigo423** [comment]
@mfuhrmann Found the issue with the sudo message, I ran it as a user and just had to run sudo before.

File: ansible/roles/horizon-provision/templates/node_xml.j2 (hunk 11 above)
[the remark this answers is not in the document]
Comment (mfuhrmann) [reply ansible/roles/horizon-provision/templates/node_xml.j2]: We use `machine_id` also for `nodeLabel`, so we are fine right now I think.

**Pull request merged**